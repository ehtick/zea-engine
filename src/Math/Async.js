import { Signal } from '../Math/Signal';

class Async {
    constructor() {
        this.__asyncCount = 0;
        this.ready = new Signal(true);

        this.incAsyncCount = function(count=1) {
            this.__asyncCount+=count;
        }.bind(this);

        this.decAsyncCount = function() {
            if (this.__asyncCount > 0){
                this.__asyncCount--;
                if (this.__asyncCount == 0) {
                    this.__asyncsCompleted();
                }
            }
        }.bind(this);

        this.__asyncsCompleted = function(){
            this.ready.emit();
        }.bind(this)
    }

    get count(){
        return this.__asyncCount;
    }

};

export {
    Async
};
// export default Async;

