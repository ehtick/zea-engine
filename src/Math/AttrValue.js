// eslint-disable-next-line camelcase
import { JSON_stringify_fixedPrecision } from './Common.js'

/**
 * Base class for values that can be stored in vertex attributes.
 * <br>
 * **Note:** These values use {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array|Float32Array} values to store their data.
 */
class AttrValue {
  /**
   * Verifies that all the numeric values inside the `this.__data` variable are a valid number.
   * Returns `false` If at least one of the values is either {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/Infinity|Infinity} or
   * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/NaN|NaN}.
   *
   * @return {boolean} - Returns the result as a boolean.
   */
  isValid() {
    for (const v of this.__data) {
      if (v == Infinity || isNaN(v)) return false
    }

    return true
  }

  /**
   * Creates a new value to wrap memory in an existing buffer.
   *
   * @param {ArrayBuffer} buffer - the buffer value.
   * @param {number} offset - the offset value.
   *
   * @todo Implement this function
   */
  static createFromFloat32Buffer(buffer, offset) {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns the number of Float32 elements used by this type. Used to calculate storage requirements for large arrays of this type.
   *
   * @todo Implement this function
   */
  static numElements() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns current Vec2 data as array. Often used to pass types to the GPU.
   *
   * @return {array} - Returns the result as an array.
   */
  asArray() {
    return this.__data
  }

  /**
   * Calls `toJSON` method and stringifies it.
   *
   * @return {string} - The return value.
   */
  toString() {
    // eslint-disable-next-line new-cap
    return JSON_stringify_fixedPrecision(this.toJSON())
  }

  /**
   * Represents the state of your class as a JSON Object.
   *
   * @todo Implement this function in the derived class.
   */
  toJSON() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }
}

export { AttrValue }
export default AttrValue
