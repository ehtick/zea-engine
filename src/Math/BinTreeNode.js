import Vec2 from './Vec2.js';

class BinTreeRectBorder {
    constructor(value, movable = false) {
        this.value = value;
        this.movable = movable;
    }
};

class BinTreeRect {
    constructor(pos, right, top) {
        this.pos = pos;
        this.right = right;
        this.top = top;
    }

    get size() {
        return new Vec2(this.right.value - this.pos.x, this.top.value - this.pos.y);
    }
};

class BinTreeNode {
    constructor(rect, rootNode = false) {
        this.rect = rect;
        this.child0 = undefined;
        this.child1 = undefined;
        this.occupied = false;
        this.rootNode = rootNode;
        if (this.rootNode) {
            this.rect.right.ownerNode = this;
            this.rect.top.ownerNode = this;
        }
    }

    insert(rectSize, closestFit = undefined) {
        if (this.child0 != undefined && this.child1 != undefined) {
            // try inserting into first child
            let res = this.child0.insert(rectSize, closestFit);
            if (res != undefined)
                return res

            // no room, insert into second.
            return this.child1.insert(rectSize, closestFit);
        } else {

            // if there's already a rectSize here, return...
            if (this.occupied) {
                return undefined;
            }
            let delta = rectSize.subtract(this.rect.size);

            // (if we're too small, return)

            // Using a custom epsilon after packing was failing on simple cases. 
            // See TestBinTree2
            let eps = 0.00001;
            if (delta.x > eps || delta.y > eps) {
                // If we can fit the cluster into the node by moving only one of the sides.
                // Check for movable sides, and if so return this as a possibility of expanding the canvas.
                if (closestFit &&
                    ((this.rect.right.movable && delta.x > 0 && delta.y <= 0) ||
                     (this.rect.top.movable && delta.y > 0 && delta.x <= 0) ||
                     (this.rect.right.movable && delta.x > 0 && this.rect.top.movable && delta.y > 0 ))) {
                    
                    // Generate a cost metric to find the node that requires 
                    // the smallest ajustment of the borders.
                    let cost = 0;
                    if (this.rect.right.movable && delta.x > 0)
                        cost += delta.x;
                    if (this.rect.top.movable && delta.y > 0)
                        cost += delta.y;

                    if (!closestFit.node || cost < closestFit.cost) {
                        closestFit.node = this;
                        closestFit.delta = delta;
                        closestFit.cost = cost;
                    }
                }
                return undefined;
            }

            // if we're just right, accept..
            if (Math.abs(delta.x) < eps && Math.abs(delta.y) < eps) {
                if(this.rect.top.movable){
                    this.sliceHorizontal(this.rect.top.value);
                    this.child0.setOccupied();
                    return this.child0;
                }
                if(this.rect.right.movable){
                    this.sliceVertical(this.rect.right.value);
                    this.child0.setOccupied();
                    return this.child0;
                }
                this.setOccupied();
                return this;
            }

            // otherwise, gotta split this node and create some kids
            if (delta.x < delta.y)
                this.sliceVertical(this.rect.pos.x + rectSize.x);
            else
                this.sliceHorizontal(this.rect.pos.y + rectSize.y);

            return this.child0.insert(rectSize, closestFit);
        }
    }

    setOccupied() {
        this.occupied = true;
    }

    sliceVertical(xDivisor) {
        let child0rect = new BinTreeRect(this.rect.pos, new BinTreeRectBorder(xDivisor), this.rect.top);
        let child1rect = new BinTreeRect(new Vec2(xDivisor, this.rect.pos.y), this.rect.right, this.rect.top);
        this.child0 = new BinTreeNode(child0rect);
        this.child1 = new BinTreeNode(child1rect);
    }

    sliceHorizontal(yDivisor) {
        let child0rect = new BinTreeRect(this.rect.pos, this.rect.right, new BinTreeRectBorder(yDivisor));
        let child1rect = new BinTreeRect(new Vec2(this.rect.pos.x, yDivisor), this.rect.right, this.rect.top);
        this.child0 = new BinTreeNode(child0rect);
        this.child1 = new BinTreeNode(child1rect);
    }

    resizeAndInsert(rectSize, delta) {
        if (delta.x > 0 && this.rect.right.movable && delta.y > 0 && this.rect.top.movable) {
            this.rect.right.value += delta.x;
            this.rect.top.value += delta.y;
            this.sliceVertical(this.rect.right.value);
            this.child0.sliceHorizontal(this.rect.top.value);
            this.child0.child0.setOccupied();
            return this.child0.child0;
        }
        else if (delta.x > 0 && this.rect.right.movable){
            this.rect.right.value += delta.x;
            this.sliceVertical(this.rect.right.value);
            let res = this.child0.insert(rectSize);
            if (res == undefined)
                console.log("We should fit now~!!");
            return res;
        }
        else if (delta.y > 0 && this.rect.top.movable) {
            this.rect.top.value += delta.y;
            this.sliceHorizontal(this.rect.top.value);
            let res = this.child0.insert(rectSize);
            if (res == undefined)
                console.log("We should fit now~!!");
            return res;
        }
        else {
            throw ("Why are we here!!");
        }
    }
};

export {
    BinTreeRect,
    BinTreeRectBorder,
    BinTreeNode
};

export default BinTreeNode;
