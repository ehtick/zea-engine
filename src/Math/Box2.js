import {
    JSON_stringify_fixedPrecision
} from './Common.js';
import {
    Vec2
} from './Vec2';
import {
    typeRegistry
} from './TypeRegistry.js';

class Box2 {
    constructor(p0 = undefined, p1 = undefined) {
        if (p0 instanceof Vec2) {
            this.p0 = p0;
        } else {
            this.p0 = new Vec2(Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY);
        }
        if (p1 instanceof Vec2) {
            this.p1 = p1;
        } else {
            this.p1 = new Vec2(Number.NEGATIVE_INFINITY, Number.NEGATIVE_INFINITY);
        }
    }


    set(p0, p1) {
        this.p0 = p0;
        this.p1 = p1;
    }

    reset() {
        this.p0.x = Number.POSITIVE_INFINITY;
        this.p1.x = Number.NEGATIVE_INFINITY;
        this.p0.y = Number.POSITIVE_INFINITY;
        this.p1.y = Number.NEGATIVE_INFINITY;
    }

    isValid() {
        return this.p0.x != Number.POSITIVE_INFINITY && this.p1.x != Number.NEGATIVE_INFINITY &&
            this.p0.y != Number.POSITIVE_INFINITY && this.p1.y != Number.NEGATIVE_INFINITY;
    }

    addPoint(point) {
        if (this.p0.x == Number.POSITIVE_INFINITY || point.x < this.p0.x)
            this.p0.x = point.x;
        if (this.p0.y == Number.POSITIVE_INFINITY || point.y < this.p0.y)
            this.p0.y = point.y;

        if (this.p1.y == Number.NEGATIVE_INFINITY || point.x > this.p1.x)
            this.p1.x = point.x;
        if (this.p1.y == Number.NEGATIVE_INFINITY || point.y > this.p1.y)
            this.p1.y = point.y;
    }

    size() {
        return this.p1.subtract(this.p0);
    }
    diagonal() {
        return this.p1.subtract(this.p0);
    }

    center() {
        let result = this.p1.subtract(this.p0);
        result.scaleInPlace(0.5);
        result.addInPlace(this.p0);
        return result;
    }


    //////////////////////////////////////////
    // Static Methods

    static create(...args) {
        return new Box2(...args);
    }

    //////////////////////////////////////////
    // Persistence

    toJSON() {
        return {
            "p0": this.p0.toJSON(),
            "p1": this.p1.toJSON()
        }
    }

    toString() {
        return JSON_stringify_fixedPrecision(this.toJSON())
    }
};

typeRegistry.registerType('Box2', Box2);

export {
    Box2
};
// export default Box2;