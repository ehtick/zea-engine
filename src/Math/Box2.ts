import { StringFunctions } from '../Utilities/StringFunctions'
import { Vec2 } from './Vec2'

/**
 * Represents a box in 2D space. Needing two Vec2 vectors describing the corners
 */
class Box2 {
  p0: Vec2
  p1: Vec2

  /**
   * Creates a Box2 object using Vec2s.
   * In case the parameters are not passed by, their values are pre-defined:
   * <br>
   * p0 is a Vec2 with {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/POSITIVE_INFINITY|`Number.POSITIVE_INFINITY`}
   * <br>
   * p1 is a Vec2 with {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/NEGATIVE_INFINITY|`Number.NEGATIVE_INFINITY`}
   *
   * @param {Vec2} p0 - A point representing the corners of a 2D box.
   * @param {Vec2} p1 - A point representing the corners of a 2D box.
   */
  constructor(p0?: Vec2, p1?: Vec2) {
    if (p0 instanceof Vec2) {
      this.p0 = p0
    } else {
      this.p0 = new Vec2(Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY)
    }
    if (p1 instanceof Vec2) {
      this.p1 = p1
    } else {
      this.p1 = new Vec2(Number.NEGATIVE_INFINITY, Number.NEGATIVE_INFINITY)
    }
  }

  /**
   * Sets both Vec2 points
   *
   * @param {Vec2} p0 - A point representing the corners of a 2D box.
   * @param {Vec2} p1 - A point representing the corners of a 2D box.
   */
  set(p0: Vec2, p1: Vec2): void {
    this.p0 = p0
    this.p1 = p1
  }

  /**
   * Resets the box2 back to an uninitialized state.
   *
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/POSITIVE_INFINITY|`Number.POSITIVE_INFINITY`}
   * and {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/NEGATIVE_INFINITY|`Number.NEGATIVE_INFINITY`}
   */
  reset(): void {
    this.p0.x = Number.POSITIVE_INFINITY
    this.p1.x = Number.NEGATIVE_INFINITY
    this.p0.y = Number.POSITIVE_INFINITY
    this.p1.y = Number.NEGATIVE_INFINITY
  }

  /**
   * Returns `true` if the box has been expanded to contain a point.
   *
   * @return {boolean} - The return value.
   */
  isValid(): boolean {
    return (
      this.p0.x != Number.POSITIVE_INFINITY &&
      this.p1.x != Number.NEGATIVE_INFINITY &&
      this.p0.y != Number.POSITIVE_INFINITY &&
      this.p1.y != Number.NEGATIVE_INFINITY
    )
  }

  /**
   * Expands the Box2 to contain the new point.
   *
   * @param {Vec2} point - A point represents the corners of a 2D box.
   */
  addPoint(point: Vec2): void {
    if (this.p0.x == Number.POSITIVE_INFINITY || point.x < this.p0.x) this.p0.x = point.x
    if (this.p0.y == Number.POSITIVE_INFINITY || point.y < this.p0.y) this.p0.y = point.y

    if (this.p1.y == Number.NEGATIVE_INFINITY || point.x > this.p1.x) this.p1.x = point.x
    if (this.p1.y == Number.NEGATIVE_INFINITY || point.y > this.p1.y) this.p1.y = point.y
  }

  /**
   * Returns the length of the diagonal of the box.
   *
   * @return {number} - Returns the distance.
   */
  size(): number {
    return this.p1.distanceTo(this.p0)
  }

  /**
   * Returns the size of a Box2 - the same as size().
   *
   * @return {Vec2} - Returns a Vec2.
   */
  diagonal(): Vec2 {
    return this.p1.subtract(this.p0)
  }

  /**
   * Returns the center point of a Box2.
   *
   * @return {Vec2} - Returns a Vec2.
   */
  center(): Vec2 {
    const result = this.p1.subtract(this.p0)
    result.scaleInPlace(0.5)
    result.addInPlace(this.p0)
    return result
  }

  // ////////////////////////////////////////
  // Persistence

  /**
   * Encodes `Box2` Class as a JSON object for persistence.
   *
   * @return {Record<string, Record<string, number>>} - The json object.
   */
  toJSON(): Record<string, Record<string, number>> {
    return {
      p0: this.p0.toJSON(),
      p1: this.p1.toJSON(),
    }
  }

  /**
   * Calls `toJSON` method and stringifies it.
   *
   * @return {string} - The return value.
   */
  toString(): string {
    // eslint-disable-next-line new-cap
    return StringFunctions.stringifyJSONWithFixedPrecision(this.toJSON())
  }
}

export { Box2 }
