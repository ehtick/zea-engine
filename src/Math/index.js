
export {
    SInt32,
    UInt32,
    Float32,
    hashStr,
    JSON_stringify_fixedPrecision
} from './Common.js';
export * from './AttrValue.js';
export * from './Vec2.js';
export * from './Vec3.js';
export * from './Vec4.js';
export * from './Color.js';
export * from './Quat.js';
export * from './Ray.js';
export * from './Mat3.js';
export * from './Mat4.js';
export * from './Xfo.js';
export * from './Box2.js';
export * from './Box3.js';
export * from './Rect2.js';
export * from './Hammersley.js';

export * from './TypeRegistry.js';
