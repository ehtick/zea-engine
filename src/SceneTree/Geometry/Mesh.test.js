import { Mesh } from './Mesh'
import { Vec2, Vec3, Box3 } from '../../Math'

describe('Mesh', () => {
  test('Check for default positions attribute.', () => {
    const mesh = new Mesh()
    expect(mesh.hasVertexAttribute('positions')).toBe(true)
  })

  test('Check for setting up a single triangle', () => {
    const mesh = new Mesh()
    const numVertices = 3
    mesh.setNumVertices(numVertices)

    const positions = mesh.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, 0, 1, 2)

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])

    expect(mesh.getVertexAttribute('positions').length).toBe(numVertices)
  })

  test('Check for setting up a single quad', () => {
    const mesh = new Mesh()
    const numVertices = 4
    mesh.setNumVertices(numVertices)

    const positions = mesh.getVertexAttribute('positions')
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)

    const normals = mesh.addVertexAttribute('normals', Vec3, 0.0)
    normals.getValueRef(0).set(-1, 0, 0)
    normals.getValueRef(1).set(-1, 0, 0)
    normals.getValueRef(2).set(-1, 0, 0)
    normals.getValueRef(3).set(-1, 0, 0)

    mesh.setFaceCounts([0, 1])
    mesh.setFaceVertexIndices(0, 0, 1, 2, 3)

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2, 3])
  })

  test('Check for setting up 2 neighboring quads', () => {
    const mesh = new Mesh()
    const numVertices = 6
    mesh.setNumVertices(numVertices)

    const positions = mesh.getVertexAttribute('positions')
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)
    positions.getValueRef(4).set(1, 1, 1)
    positions.getValueRef(5).set(1, 1, -1)

    mesh.setFaceCounts([0, 2])
    mesh.setFaceVertexIndices(0, 0, 1, 2, 3)
    mesh.setFaceVertexIndices(1, 3, 2, 4, 5)

    const normals = mesh.addVertexAttribute('normals', Vec3, 0.0)

    // The first face has normals along the -X axis
    normals.setFaceVertexValue(0, 0, new Vec3(-1, 0, 0))
    normals.setFaceVertexValue(0, 1, new Vec3(-1, 0, 0))
    normals.setFaceVertexValue(0, 2, new Vec3(-1, 0, 0))
    normals.setFaceVertexValue(0, 3, new Vec3(-1, 0, 0))

    // The second face has normals along the -Y axis
    normals.setFaceVertexValue(1, 0, new Vec3(0, -1, 0))
    normals.setFaceVertexValue(1, 1, new Vec3(0, -1, 0))
    normals.setFaceVertexValue(1, 2, new Vec3(0, -1, 0))
    normals.setFaceVertexValue(1, 3, new Vec3(0, -1, 0))

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2, 3])
    expect(mesh.getFaceVertexIndices(1)).toEqual([3, 2, 4, 5])
  })

  test('Check for setting up 2 neighboring quads', () => {
    const mesh = new Mesh()
    const numVertices = 6
    mesh.setNumVertices(numVertices)

    const positions = mesh.getVertexAttribute('positions')
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)
    positions.getValueRef(4).set(1, 1, 1)
    positions.getValueRef(5).set(1, 1, -1)

    mesh.setFaceCounts([0, 1])
    mesh.setFaceVertexIndices(0, 0, 1, 2, 3)
    mesh.setFaceVertexIndices(1, 3, 2, 4, 5)

    mesh.computeVertexNormals()

    expect(JSON.stringify(mesh.toJSON())).toMatchSnapshot()
  })

  test.skip('Check resizing bigger the face indices.', () => {
    const mesh = new Mesh()
    const numVertices = 4
    mesh.setNumVertices(numVertices)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, 0, 1, 2)

    mesh.setFaceCounts([2])
    mesh.setFaceVertexIndices(1, 2, 1, 3)

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
    expect(mesh.getFaceVertexIndices(1)).toEqual([2, 1, 3])
  })

  test.skip('Check resizing smaller the face indices.', () => {
    const mesh = new Mesh()
    const numVertices = 4
    mesh.setNumVertices(numVertices)

    mesh.setFaceCounts([2])
    mesh.setFaceVertexIndices(0, 0, 1, 2)
    mesh.setFaceVertexIndices(1, 2, 1, 3)

    mesh.setFaceCounts([1])

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
  })

  test('Check generated buffers', () => {
    const mesh = new Mesh()
    const numVertices = 3
    mesh.setNumVertices(numVertices)

    const positions = mesh.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, 0, 1, 2)

    expect(mesh.genBuffers()).toEqual({
      attrBuffers: {
        positions: {
          count: 3,
          dataType: Vec3,
          dimension: 3,
          normalized: false,
          values: new Float32Array([1, 2, 3, -1, -2, -3, 2, 1, -3]),
        },
      },
      indices: new Uint8Array([0, 1, 2]),
      numRenderVerts: 3,
      numVertices: 3,
    })
  })

  test('saves to JSON (serialization).', () => {
    const mesh = new Mesh()
    const numVertices = 3
    mesh.setNumVertices(numVertices)

    const positions = mesh.getVertexAttribute('positions')
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, 0, 1, 2)

    expect(JSON.stringify(mesh.toJSON())).toMatchSnapshot()
  })

  test.skip('loads from JSON (serialization).', () => {
    const mesh = new Mesh()
    const input = {
      type: 'Mesh',
      numVertices: 3,
      vertexAttributes: {
        positions: {
          data: [1, 2, 3, -1, -2, -3, 2, 1, -3],
          dataType: 'Vec3',
          defaultValue: 0,
          length: 3,
          splits: {},
          splitValues: [],
        },
      },
      faceVertexIndices: [0, 1, 2],
    }
    mesh.fromJSON(input)

    expect(mesh.getVertexAttribute('positions').length).toBe(3)
    expect(mesh.getFaceCounts()).toEqual([2])
  })
})
