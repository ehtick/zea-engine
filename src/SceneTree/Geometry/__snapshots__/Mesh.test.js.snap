// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mesh Check computing normals on 2 neighboring quads 1`] = `"{\\"type\\":\\"Mesh\\",\\"numVertices\\":6,\\"vertexAttributes\\":{\\"positions\\":{\\"data\\":[0,-1,-1,0,-1,1,0,1,1,0,1,-1,1,1,1,1,1,-1],\\"dataType\\":\\"Vec3\\",\\"defaultValue\\":0,\\"length\\":6,\\"splits\\":{},\\"splitValues\\":[]},\\"normals\\":{\\"data\\":[-1,0,0,-1,0,0,-1,0,0,0,1,0,0,1,0,0,1,0],\\"dataType\\":\\"Vec3\\",\\"defaultValue\\":1.7976931348623157e+308,\\"length\\":6,\\"splits\\":{\\"2\\":{\\"1\\":0},\\"3\\":{\\"0\\":1}},\\"splitValues\\":[{\\"x\\":0,\\"y\\":1,\\"z\\":0},{\\"x\\":-1,\\"y\\":0,\\"z\\":0}]}},\\"faceCounts\\":[0,2,0],\\"faceVertexIndices\\":[0,1,2,3,3,2,4,5]}"`;

exports[`Mesh saves to JSON (serialization). 1`] = `"{\\"type\\":\\"Mesh\\",\\"numVertices\\":3,\\"vertexAttributes\\":{\\"positions\\":{\\"data\\":[1,2,3,-1,-2,-3,2,1,-3],\\"dataType\\":\\"Vec3\\",\\"defaultValue\\":0,\\"length\\":3,\\"splits\\":{},\\"splitValues\\":[]}},\\"faceCounts\\":[1],\\"faceVertexIndices\\":[0,1,2]}"`;
